coil-network-core/src/commonMain/kotlin/coil3/network/NetworkFetcher.kt

```diff
@@ -69,37 +69,51 @@ class NetworkFetcher(
                 }
             }
 
+            // Prevent executing requests on the main thread that could block due to a
+            // networking operation.
+            if (options.networkCachePolicy.readEnabled) {
+                assertNotOnMainThread()
+            }
+
             // Slow path: fetch the image from the network.
             val networkRequest = readResult?.request ?: newRequest()
-            var fetchResult = executeNetworkRequest(networkRequest) { response ->
+            var fetchResult = networkClient.value.executeRequest(networkRequest) { networkResponse ->
                 // Write the response to the disk cache then open a new snapshot.
-                snapshot = writeToDiskCache(snapshot, cacheResponse, networkRequest, response)
+                snapshot = writeToDiskCache(snapshot, cacheResponse, networkRequest, networkResponse)
+
+                // Throw if the network response returns a non-200 (or 304) response.
+                if (networkResponse.code !in 200 until 300 &&
```

Comment:
This is the main part of the fix. Previously we were throwing before we checked the `CacheControlStrategy` and before `writeToDiskCache`.